kernel/thread.c

```diff
@@ -99,10 +99,11 @@ void _impl_k_busy_wait(u32_t usec_to_wait)
 {
 #if !defined(CONFIG_ARCH_HAS_CUSTOM_BUSY_WAIT)
 	/* use 64-bit math to prevent overflow when multiplying */
+	u64_t usecs = USEC_PER_SEC;
 	u32_t cycles_to_wait = (u32_t)(
-		(u64_t)usec_to_wait *
-		(u64_t)sys_clock_hw_cycles_per_sec() /
-		(u64_t)USEC_PER_SEC
+		usec_to_wait *
+		sys_clock_hw_cycles_per_sec() /
+		usecs
```

Comment:
This still looks wrong to me.  That is going to evaluate as:

    (usec_to_wait * s_c_h_c_p_s() )  / usecs

The multiplication is still being done in 32 bits and will overflow where the original expression wasn't.

I also still don't understant why MISRA disallows doing math in higher precision.  Does it issue the same warning if you want to extract a bitfield or take a char variable and use it as a multiplier on an int?

Reply:
arghh, indeed. I'll remove this commit from this pr, since it will require some extra work.